crux-core/src/crux/query.clj

```diff
@@ -1122,36 +1123,37 @@
          stats (idx/read-meta kv :crux.kv/stats)]
      (log/debug :query (cio/pr-edn-str q))
      (validate-args args)
-     (let [rule-name->rules (rule-name->rules rules)
-           entity-as-of-idx (idx/new-entity-as-of-index snapshot valid-time transact-time)
-           entity-as-of-idx (if *with-entities-cache?*
-                              (lru/new-cached-index entity-as-of-idx default-entity-cache-size)
-                              entity-as-of-idx)
-           db (assoc db :entity-as-of-idx entity-as-of-idx)
-           {:keys [n-ary-join
-                   var->bindings]} (build-sub-query snapshot db where args rule-name->rules stats)]
-       (doseq [var find
-               :when (not (contains? var->bindings var))]
-         (throw (IllegalArgumentException.
-                 (str "Find refers to unknown variable: " var))))
-       (doseq [{:keys [var]} order-by
-               :when (not (contains? var->bindings var))]
-         (throw (IllegalArgumentException.
-                 (str "Order by refers to unknown variable: " var))))
-       (doseq [{:keys [var]} order-by
-               :when (not (some #{var} find))]
-         (throw (IllegalArgumentException.
-                  (str "Order by requires a var from :find. unreturned var: " var))))
-       (cond->> (for [[join-keys join-results] (idx/layered-idx->seq n-ary-join)
-                      :let [bound-result-tuple (for [var find]
-                                                 (bound-result-for-var snapshot object-store var->bindings join-keys join-results var))]]
-                  (if full-results?
-                    (build-full-results db snapshot bound-result-tuple)
-                    (mapv #(.value ^BoundResult %) bound-result-tuple)))
-         order-by (cio/external-sort (order-by-comparator find order-by))
-         (or offset limit) dedupe
-         offset (drop offset)
-         limit (take limit))))))
+     (with-open [i (kv/new-iterator snapshot)]
```

Comment:
This should not use with-open here as this is what returns the lazy seq and this iterator needs to be open as long as the snapshot is, so we should just create it (like before) and let it live until the snapshot gets closed, like most iterators created by the query engine.

Reply:
thanks - yep, with-open removed and inlined binding. cheers